cpp/merkletree/merkle_tree.cc

```diff
@@ -289,3 +289,99 @@ void MerkleTree::AddLevel() {
 size_t MerkleTree::LazyLevelCount() const {
   return tree_.size();
 }
+
+MutableMerkleTree::MutableMerkleTree(unique_ptr<SerialHasher> hasher)
+    : MerkleTree(move(hasher)) {}
+
+MutableMerkleTree::~MutableMerkleTree() {}
+
+bool MutableMerkleTree::UpdateLeaf(size_t leaf, const string& data) {
+  // Optimized for success: hash computed before validating the leaf's position.
+  return UpdateLeafHash(leaf, treehasher_.HashLeaf(data));
+}
+
+bool MutableMerkleTree::UpdateLeafHash(size_t leaf, const string& hash) {
+  if (leaf == 0 || leaf > LeafCount()) return false;
+
+  // Update the leaf node.
+  size_t child = leaf - 1;
+  tree_[0].replace(treehasher_.DigestSize() * child, treehasher_.DigestSize(),
+                   hash);
+
+  if (leaf > leaves_processed_) return true;
+
+  // Update the parents chain, level by level.
+  size_t child_level = 0;
+  size_t parent = MerkleTreeMath::Parent(child);
+  string parent_hash;
+  while (child) {
+    if (MerkleTreeMath::IsRightChild(child)) {
+      parent_hash = treehasher_.HashChildren(Node(child_level, child - 1),
+                                             Node(child_level, child));
+    } else if (child < NodeCount(child_level) - 1) {
+      parent_hash = treehasher_.HashChildren(Node(child_level, child),
+                                             Node(child_level, child + 1));
+    } else {
+      // Propagate the "dummy" node.
+      parent_hash = Node(child_level, child);
+    }
+
+    tree_[child_level + 1].replace(treehasher_.DigestSize() * parent,
+                                   treehasher_.DigestSize(), parent_hash);
+
+    child = parent;
+    parent = MerkleTreeMath::Parent(parent);
+    ++child_level;
+  }
+
+  return true;
+}
+
+bool MutableMerkleTree::Truncate(size_t leaf) {
+  if (leaf > LeafCount()) return false;
+
+  if (leaf == 0) {
+    if (leaves_processed_ > 0) {
```

Comment:
I think the case is truncating the tree to size zero when nothing has been propagated through the lazy evaluation. I'm now suspicious that it doesn't clear the tree in this case because you might have added stuff that has not yet been propagated.